ext/drx/drx_buf.c

```diff
@@ -648,6 +650,116 @@ drx_buf_insert_buf_store(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
     }
 }
 
+#ifndef AARCH64
+/* FIXME i#1569: NYI on AArch64 */
+static void
+insert_load(void *drcontext, instrlist_t *ilist, instr_t *where,
+                    reg_id_t dst, reg_id_t src, opnd_size_t opsz)
+{
+    switch (opsz) {
+    case OPSZ_1:
+        MINSERT(ilist, where, XINST_CREATE_load_1byte
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_2:
+        MINSERT(ilist, where, XINST_CREATE_load_2bytes
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_4:
+#if defined(X86_64) || defined(AARCH64)
+    case OPSZ_8:
+#endif
+        MINSERT(ilist, where, XINST_CREATE_load
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    default:
+        DR_ASSERT(false);
+        break;
+    }
+}
+
+/* Performs a drx_buf-compatible memcpy which handles its own fault.
+ * Note that on a fault we simply reset the buffer pointer with no partial write.
+ */
+static void
+safe_memcpy(drx_buf_t *buf, void *src, size_t len)
+{
+    void *drcontext = dr_get_current_drcontext();
+    per_thread_t *data = drmgr_get_tls_field(drcontext, buf->tls_idx);
+    byte *cli_ptr = BUF_PTR(data->seg_base, buf->tls_offs);
+    size_t written;
+
+    /* try to perform a safe memcpy */
+    dr_safe_write(cli_ptr, len, src, &written);
+    if (written != len) {
```

Comment:
Safer to test dr_safe_write return value too in case it doesn't write output param on outright failure